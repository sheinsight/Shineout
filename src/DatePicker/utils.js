import addDays from 'date-fns/addDays'
import addMonths from 'date-fns/addMonths'
import addSeconds from 'date-fns/addSeconds'
import addYears from 'date-fns/addYears'
import compareAsc from 'date-fns/compareAsc'
import endOfMonth from 'date-fns/endOfMonth'
import endOfWeek from 'date-fns/endOfWeek'
import format from 'date-fns/format'
import isSameDay from 'date-fns/isSameDay'
import isSameMonth from 'date-fns/isSameMonth'
import isSameWeek from 'date-fns/isSameWeek'
import isValid from 'date-fns/isValid'
import parse from 'date-fns/parse'
import startOfMonth from 'date-fns/startOfMonth'
import startOfWeek from 'date-fns/startOfWeek'
import toDate from 'date-fns/toDate'

const TIME_FORMAT = 'HH:mm:ss'

function getDaysOfMonth(dirtyDate) {
  const date = toDate(dirtyDate)
  const end = endOfWeek(endOfMonth(date))
  let current = startOfWeek(startOfMonth(date))
  current.setHours(dirtyDate.getHours())
  current.setMinutes(dirtyDate.getMinutes())
  current.setSeconds(dirtyDate.getSeconds())

  const days = []

  while (current.getTime() < end.getTime()) {
    days.push(current)
    current = addDays(current, 1)
  }

  return days
}

function isInvalid(date) {
  // eslint-disable-next-line
  return isNaN(date)
}

function toDateWithFormat(dirtyDate, fmt, def) {
  let date
  if (typeof dirtyDate === 'string') date = parse(dirtyDate, fmt, new Date())
  else date = toDate(dirtyDate)

  if (isInvalid(date)) date = toDate(dirtyDate)
  if (isInvalid(date)) date = def

  return date
}

function compareMonth(dateLeft, dateRight, pad = 0) {
  if (!dateLeft || !dateRight) return 0
  const left = new Date(dateLeft.getFullYear(), dateLeft.getMonth(), 1)
  const right = new Date(dateRight.getFullYear(), dateRight.getMonth() + pad, 1)
  return compareAsc(left, right)
}

function newDate() {
  const date = new Date()
  return new Date(date.getFullYear(), date.getMonth(), date.getDate())
}

function cloneTime(date, old, fmt) {
  old = toDateWithFormat(old, fmt)
  if (isInvalid(old)) return date

  date.setHours(old.getHours())
  date.setMinutes(old.getMinutes())
  date.setSeconds(old.getSeconds())

  return date
}

function formatDateWithDefaultTime(date, defaultTime, fmt) {
  if (!defaultTime) return date
  const dateHMS = toDateWithFormat(defaultTime, TIME_FORMAT)
  if (isInvalid(dateHMS)) return date

  const nDate = cloneTime(date, defaultTime, TIME_FORMAT)
  return format(nDate, fmt)
}

export default {
  addDays,
  addMonths,
  addYears,
  addSeconds,
  cloneTime,
  compareAsc,
  compareMonth,
  getDaysOfMonth,
  format,
  isInvalid,
  isSameDay,
  isSameMonth,
  isSameWeek,
  isValid,
  newDate,
  parse,
  toDate,
  toDateWithFormat,
  formatDateWithDefaultTime,
  TIME_FORMAT,
}
