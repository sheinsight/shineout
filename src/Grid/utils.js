import config from '../config'

const CACHES = {}
const RESPONSIVE = {
  sm: '568',
  md: '768',
  lg: '992',
  xl: '1200',
}
const GridClassName = `${config.prefix}-grid`
const GridFullClassName = `${config.prefix}-grid-full`
const defaultResponsive = 'md'

function createStyle(text, id) {
  let style = document.head.querySelector(`#${id}`)
  if (style) {
    return
  }

  style = document.createElement('style')
  style.type = 'text/css'
  style.id = id
  style.innerHTML = text
  document.head.appendChild(style)
}

function generateGrid(width, className, responsive) {
  const minWidth = RESPONSIVE[responsive]
  const text = `@media screen and (min-width: ${minWidth}px) { .${GridClassName}.${className}{width: ${width}%} }`
  createStyle(text, className)
}

function generateOffset(width, className, responsive) {
  const minWidth = RESPONSIVE[responsive]
  const text = `@media screen and (min-width: ${minWidth}px) { .${className}{margin-left: ${width}%} }`
  createStyle(text, className)
}

function generate(w, type, res) {
  let width = w
  const responsive = res || defaultResponsive

  if (!width || width <= 0) {
    return ''
  }

  if (width > 1) {
    width = 1
  }
  width = (width * 100).toFixed(4)
  width = width.substr(0, width.length - 1)

  const className = `${config.prefix}-${type}-${responsive}-${width.replace('.', '-')}`

  if (!CACHES[className]) {
    if (type === 'grid') {
      generateGrid(width, className, responsive)
    } else {
      generateOffset(width, className, responsive)
    }
    CACHES[className] = true
  }
  return className
}

export function getGrid(opt) {
  let options = opt
  if (!options) {
    return ''
  }
  if (typeof options === 'number') {
    options = { width: options }
  }

  const { width, offset, responsive } = options
  const gridClass = generate(width, 'grid', responsive)
  const offsetClass = generate(offset, 'offset', responsive)

  return `${GridClassName} ${GridFullClassName} ${gridClass} ${offsetClass}`
}

function init() {
  const text = []

  text.push(`
.${GridClassName} {
  position: relative;
  display: inline-block;
  zoom: 1;
  letter-spacing: normal;
  word-spacing: normal;
  vertical-align: top;
  text-rendering: auto;
  box-sizing: border-box;
}`)

  text.push(`.${GridFullClassName}{width:100%}`)
  createStyle(text.join(''), GridClassName)
}

init()
