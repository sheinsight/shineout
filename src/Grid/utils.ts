import config from '../config'

const CACHES: {
  [className: string]: boolean
} = {}
const RESPONSIVE = {
  sm: '568',
  md: '768',
  lg: '992',
  xl: '1200',
}
type Responsive = keyof typeof RESPONSIVE

const GridClassName = `${config.prefix}-grid`
const GridFullClassName = `${config.prefix}-grid-full`
const defaultResponsive = 'md'

function createStyle(text: string, id: string) {
  let style = document.head.querySelector(`#${id}`) as Element & { type: string }
  if (style) {
    return
  }

  style = document.createElement('style')
  style.type = 'text/css'
  style.id = id
  style.innerHTML = text
  document.head.appendChild(style)
}

function generateGrid(width: number | string, className: string, responsive: Responsive) {
  const minWidth = RESPONSIVE[responsive]
  const text = `@media screen and (min-width: ${minWidth}px) { .${className}{width: ${width}%} }`
  createStyle(text, className)
}

function generateOffset(width: number | string, className: string, responsive: Responsive) {
  const minWidth = RESPONSIVE[responsive]
  const text = `@media screen and (min-width: ${minWidth}px) { .${className}{margin-left: ${width}%} }`
  createStyle(text, className)
}

function generate(w: number | string, type: 'grid' | 'offset', res: Responsive) {
  let width = w
  const responsive = res || defaultResponsive

  if (!width || width <= 0) {
    return ''
  }

  if (width > 1) {
    width = 1
  }
  width = ((width as number) * 100).toFixed(4)
  width = width.substr(0, width.length - 1)

  const className = `${config.prefix}-${type}-${responsive}-${width.replace('.', '-')}`
  if (!CACHES[className]) {
    if (type === 'grid') {
      generateGrid(width, className, responsive)
    } else {
      generateOffset(width, className, responsive)
    }
    CACHES[className] = true
  }
  return className
}

export function getGrid(
  opt?:
    | number
    | {
        width?: number
        offset?: number | string
        responsive?: Responsive
      }
) {
  let options = opt
  if (!options) {
    return ''
  }
  if (typeof options === 'number') {
    options = { width: options }
  }

  const { width, offset, responsive } = options
  const gridClass = generate(width!, 'grid', responsive!)
  const offsetClass = generate(offset!, 'offset', responsive!)

  return `${GridClassName} ${GridFullClassName} ${gridClass} ${offsetClass}`
}

function init() {
  const text = []

  text.push(`
.${GridClassName} {
  position: relative;
  display: inline-block;
  zoom: 1;
  letter-spacing: normal;
  word-spacing: normal;
  vertical-align: top;
  text-rendering: auto;
  box-sizing: border-box;
}`)

  text.push(`.${GridFullClassName}{width:100%}`)
  createStyle(text.join(''), GridClassName)
}

init()
