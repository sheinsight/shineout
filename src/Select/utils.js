import { isFunc } from '../utils/is'

export const getCustomList = (list, renderOptionList, loading) => {
  if (renderOptionList && isFunc(renderOptionList)) return renderOptionList(list, { loading })
  return list
}
