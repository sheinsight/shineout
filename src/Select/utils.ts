import React, { ReactElement } from 'react'
import { isFunc } from '../utils/is'

export const getCustomList = (
  list: React.ReactNode,
  renderOptionList?: (list: React.ReactNode, info: { loading: React.ReactNode }) => ReactElement,
  loading?: React.ReactNode
) => {
  if (renderOptionList && isFunc(renderOptionList)) return renderOptionList(list, { loading })
  return list
}
