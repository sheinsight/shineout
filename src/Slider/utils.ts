import { toPrecision } from '../utils/numbers'

export function value2per(value: number, scale: number[]) {
  const range = scale.length - 1
  let ps = 0
  scale.forEach((s, i) => {
    if (value > s) ps = i
  })

  // end
  if (ps >= range) return 1

  const min = scale[ps]
  const max = scale[ps + 1]

  return (ps + (value - min) / (max - min)) / range
}

export function per2value(per: number, scale: number[], step = 1) {
  const range = scale.length - 1

  if (step === 0) return scale[Math.round(per * range)]

  if (per >= 1) return scale[range]
  const ps = Math.floor(per * range)
  const min = scale[ps]
  const max = scale[ps + 1]
  const count = (max - min) / step
  const sper = (per - ps / range) * range

  return toPrecision(min + Math.round(sper * count) * step)
}
