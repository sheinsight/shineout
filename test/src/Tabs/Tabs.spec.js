import React from 'react'
import { Tabs } from 'shineout'
import { mount } from 'enzyme'
import { Link, BrowserRouter } from 'react-router-dom'
import { baseTest, delay } from '../../utils'
import TabsActive from '../../../site/pages/components/Tabs/example-05-active'
import TabsShapeLine from '../../../site/pages/components/Tabs/example-06-shape-line'
import TabsShapeButton from '../../../site/pages/components/Tabs/example-07-shape-button'
import TabsCollapsible from '../../../site/pages/components/Tabs/example-11-collapsible'

/* global SO_PREFIX */
describe('Tabs[Active]', () => {
  test('should active controlled', () => {
    const wrapper = mount(<TabsActive />)
    wrapper.find(`.${SO_PREFIX}-checkinput`).forEach(radio => {
      const text = radio
        .find(`.${SO_PREFIX}-checkinput-desc`)
        .first()
        .text()

      // simulate change
      radio
        .find(`input[type="radio"]`)
        .first()
        .simulate('change', () => ({
          target: {
            checked: true,
          },
        }))

      wrapper.update()

      const tabText = wrapper
        .find(`.${SO_PREFIX}-tabs-active`)
        .first()
        .text()

      expect(text[0].toUpperCase() + text.substring(1) === tabText).toBe(true)
    })
  })
})

describe('Tabs[Shape-Line]', () => {
  test('should render line class', () => {
    const lineWrapper = mount(<TabsShapeLine />)
    expect(lineWrapper.find(`.${SO_PREFIX}-tabs-line`)).toHaveLength(1)
    const buttonWrapper = mount(<TabsShapeButton />)
    expect(buttonWrapper.find(`.${SO_PREFIX}-tabs-button`)).toHaveLength(1)
  })
})

describe('Tabs[align]', () => {
  test('should render align class', () => {
    ;['left', 'right', 'vertical-left', 'vertical-right'].forEach(align => {
      const wrapper = mount(
        <Tabs defaultActive={1} align={align}>
          <Tabs.Panel tab="Home">Test</Tabs.Panel>
        </Tabs>
      )
      expect(wrapper.find(`.${SO_PREFIX}-tabs-align-${align}`)).toHaveLength(1)
    })
  })
})

describe('Tabs[collapsible]', () => {
  test('should render expand button', () => {
    const wrapper = mount(<TabsCollapsible />)
    document.body.innerHTML += wrapper.html()
    expect(wrapper.find(`.${SO_PREFIX}-tabs-header span.${SO_PREFIX}-tabs-indicator`)).toHaveLength(1)

    // default expand
    expect(wrapper.find(`div.${SO_PREFIX}-hidable-collapse`).hasClass(`${SO_PREFIX}-hidable-show`)).toBeTruthy()
    wrapper.find(`.${SO_PREFIX}-tabs-header span.${SO_PREFIX}-tabs-indicator`).simulate('click')
    expect(wrapper.find(`div.${SO_PREFIX}-hidable-collapse`).hasClass(`${SO_PREFIX}-hidable-show`)).toBeFalsy()
  })

  test('should set collapsible', async () => {
    jest.useRealTimers()
    const wrapper = mount(
      <Tabs defaultActive={1} collapsible>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
      </Tabs>
    )
    const handleCollapse = jest.fn()
    wrapper.instance().handleCollapse = handleCollapse
    wrapper.find(`.${SO_PREFIX}-tabs-tab`).forEach(i => {
      i.simulate('click')
    })
    await delay(200)
    expect(handleCollapse).toBeCalledTimes(2)
  })
})

describe('Tabs[extra]', () => {
  test('should render a extra button', () => {
    const wrapper = mount(
      <Tabs defaultActive={1}>
        <Tabs.Panel tab="Home">Test</Tabs.Panel>
      </Tabs>
    )
    expect(wrapper.find(`.${SO_PREFIX}-tabs-extra`)).toHaveLength(0)

    const wrapper1 = mount(
      <Tabs defaultActive={1} tabBarExtraContent="niconiconi">
        <Tabs.Panel tab="Home">Test</Tabs.Panel>
      </Tabs>
    )
    expect(wrapper1.find(`.${SO_PREFIX}-tabs-extra`)).toHaveLength(1)
  })
})

describe('Tabs[autoFill]', () => {
  test('should set autoFill', () => {
    const wrapper = mount(
      <div style={{ height: 500 }}>
        <Tabs defaultActive={1} autoFill>
          <Tabs.Panel tab="Home">Test</Tabs.Panel>
        </Tabs>
      </div>
    )

    expect(wrapper.find(`.${SO_PREFIX}-tabs`).hasClass(`${SO_PREFIX}-tabs-auto-fill`)).toBe(true)
  })
})

describe('Tabs[background]', () => {
  test('should set background', () => {
    const background = 'blue'
    const wrapper = mount(
      <Tabs defaultActive={1} background={background}>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )

    expect(wrapper.find(`.${SO_PREFIX}-tabs-active`).getDOMNode().style.background).toBe(background)
  })
})

describe('Tabs[border]', () => {
  test('should set border (this attribute will change border-color)', () => {
    const border = 'blue'
    const wrapper = mount(
      <Tabs defaultActive={1} border={border}>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )
    expect(wrapper.find(`.${SO_PREFIX}-tabs-hr`).getDOMNode().style.borderColor).toBe(border)
  })
})

describe('Tabs[base]', () => {
  test('should custom style and className', () => {
    baseTest(Tabs, `.${SO_PREFIX}-tabs`)
  })
})

describe('Tabs[color]', () => {
  test('should set color when shape is card', () => {
    const color = 'red'
    const wrapper = mount(
      <Tabs shape="card" defaultActive={1} color={color}>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )
    expect(wrapper.find(`.${SO_PREFIX}-tabs-active`).getDOMNode().style.color).toBe(color)
  })
})

describe('Tabs[defaultActive]', () => {
  test('should set defaultActive', () => {
    const defaultActive = 2
    const wrapper = mount(
      <Tabs defaultActive={defaultActive}>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )
    expect(
      wrapper
        .find(`.${SO_PREFIX}-tabs-tab`)
        .last()
        .hasClass(`${SO_PREFIX}-tabs-active`)
    ).toBe(true)
  })
})

describe('Tabs[defaultCollapsed]', () => {
  test('should set defaultCollapsed', () => {
    jest.useRealTimers()
    const wrapper = mount(
      <Tabs defaultActive={1} collapsible defaultCollapsed>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )
    expect(wrapper.find(`.${SO_PREFIX}-list.${SO_PREFIX}-hidable-collapse`).length).toBe(1)
  })
})

describe('Tabs[hideSplit]', () => {
  test('should set hideSplit', () => {
    const wrapper = mount(
      <Tabs defaultActive={1} hideSplit>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )
    expect(wrapper.find(`.${SO_PREFIX}-tabs-hr`).length).toBe(0)
  })
})

describe('Tabs[inactiveBackground]', () => {
  test('should set inactiveBackground', () => {
    const inactiveBackground = 'green'
    const wrapper = mount(
      <Tabs defaultActive={1} inactiveBackground={inactiveBackground}>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )
    expect(wrapper.find(`.${SO_PREFIX}-tabs-tab:not(.${SO_PREFIX}-tabs-active)`).length).toBe(2)
    wrapper.find(`.${SO_PREFIX}-tabs-tab:not(.${SO_PREFIX}-tabs-active)`).forEach(i => {
      expect(i.getDOMNode().style.background).toBe(inactiveBackground)
    })
    expect(wrapper.find(`.${SO_PREFIX}-tabs-active`).getDOMNode().style.background).toBe('')
  })
})

describe('Tabs[lazy]', () => {
  test('should set lazy when lazy is True', () => {
    const wrapper = mount(
      <Tabs defaultActive={1} lazy>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )
    expect(wrapper.find(`.${SO_PREFIX}-tabs-panel`).length).toBe(1)
    expect(wrapper.find(`.${SO_PREFIX}-tabs-tab`).length).toBe(3)
    wrapper.find(`.${SO_PREFIX}-tabs-tab`).forEach(i => {
      i.simulate('click')
    })
    expect(wrapper.find(`.${SO_PREFIX}-tabs-panel`).length).toBe(wrapper.find(`.${SO_PREFIX}-tabs-tab`).length)
  })

  test('should set lazy when lazy is False', () => {
    const wrapper = mount(
      <Tabs defaultActive={1} lazy={false}>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )
    expect(wrapper.find(`.${SO_PREFIX}-tabs-panel`).length).toBe(wrapper.find(`.${SO_PREFIX}-tabs-tab`).length)
  })
})

describe('Tabs[onChange]', () => {
  test('should set onChange', () => {
    const handleChange = jest.fn()
    const wrapper = mount(
      <Tabs defaultActive={1} onChange={handleChange}>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )
    wrapper.find(`.${SO_PREFIX}-tabs-tab`).forEach(i => {
      i.simulate('click')
    })
    expect(handleChange).toBeCalledTimes(wrapper.find(`.${SO_PREFIX}-tabs-tab`).length)
  })
})

describe('Tabs[shape]', () => {
  test('should set onChange', () => {
    const shape = ['card', 'line', 'button', 'bordered', 'dash']
    shape.forEach(i => {
      const wrapper = mount(
        <Tabs defaultActive={1} shape={i}>
          <Tabs.Panel tab="A">Test</Tabs.Panel>
          <Tabs.Panel tab="B">Test</Tabs.Panel>
          <Tabs.Panel tab="C">Test</Tabs.Panel>
        </Tabs>
      )
      expect(wrapper.find(`.${SO_PREFIX}-tabs`).hasClass(`${SO_PREFIX}-tabs-${i}`)).toBe(true)
    })
  })
})

// this attribute will be test by e2e
// describe('Tabs[sticky]', () => {
//   test('should set sticky', () => {})
// })

describe('Tabs[switchToTop]', () => {
  test('should set switchToTop when sticky is True', () => {
    const switchToTop = jest.fn()

    const wrapper = mount(
      <Tabs defaultActive={1} sticky switchToTop>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )
    // set component state to simulate switchToTop
    wrapper.instance().container.scrollIntoView = switchToTop
    wrapper.instance().sticky = true
    wrapper
      .find(`.${SO_PREFIX}-tabs-tab`)
      .first()
      .simulate('click')
    expect(switchToTop).toBeCalled()
  })
})

describe('Tabs[tabBarExtraContent]', () => {
  test('should set tabBarExtraContent and typeof tabBarExtraContent is String', () => {
    const tabBarExtraContent = 'Hello'
    const wrapper = mount(
      <Tabs defaultActive={1} tabBarExtraContent={tabBarExtraContent}>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )

    expect(wrapper.find(`.${SO_PREFIX}-tabs-extra`).length).toBe(1)
    expect(wrapper.find(`.${SO_PREFIX}-tabs-extra`).text()).toBe(tabBarExtraContent)
  })

  test('should set tabBarExtraContent and typeof tabBarExtraContent is ReactNode', () => {
    const tabBarExtraContent = <div className="hello">Hello</div>
    const wrapper = mount(
      <Tabs defaultActive={1} tabBarExtraContent={tabBarExtraContent}>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )

    expect(wrapper.find(`.${SO_PREFIX}-tabs-extra`).length).toBe(1)
    expect(
      wrapper
        .find(`.${SO_PREFIX}-tabs-extra`)
        .children()
        .hasClass('hello')
    ).toBe(true)
  })
})

describe('Tabs[tabBarStyle]', () => {
  test('should set tabBarStyle', () => {
    const tabBarStyle = 'red'
    const wrapper = mount(
      <Tabs defaultActive={1} tabBarStyle={{ color: tabBarStyle }}>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C">Test</Tabs.Panel>
      </Tabs>
    )
    expect(wrapper.find(`.${SO_PREFIX}-tabs-header`).getDOMNode().style.color).toBe(tabBarStyle)
  })
})

describe('Tabs.Panel[background]', () => {
  test('should set Tabs.Panel background', () => {
    const pannel = [
      {
        tab: 'A',
        background: 'red',
      },
      {
        tab: 'B',
        background: 'blue',
      },
      {
        tab: 'C',
        background: 'green',
      },
    ]
    const wrapper = mount(
      <Tabs defaultActive={1}>
        {pannel.map(i => (
          <Tabs.Panel key={i} tab={i.tab} background={i.background}>
            Test
          </Tabs.Panel>
        ))}
      </Tabs>
    )

    wrapper.find(`.${SO_PREFIX}-tabs-tab`).forEach((i, index) => {
      i.simulate('click')
      wrapper.update()
      expect(
        wrapper
          .find(`.${SO_PREFIX}-tabs-panel`)
          .at(index)
          .getDOMNode().style.background
      ).toBe(pannel[index].background)
    })
  })
})

describe('Tabs.Panel[border]', () => {
  test('should set Tabs.Panel border', () => {
    const pannel = [
      {
        tab: 'A',
        border: 'red',
      },
      {
        tab: 'B',
        border: 'blue',
      },
      {
        tab: 'C',
        border: 'green',
      },
    ]
    const wrapper = mount(
      <Tabs defaultActive={1}>
        {pannel.map(i => (
          <Tabs.Panel key={i} tab={i.tab} border={i.border}>
            Test
          </Tabs.Panel>
        ))}
      </Tabs>
    )

    wrapper.find(`.${SO_PREFIX}-tabs-tab`).forEach((i, index) => {
      i.simulate('click')
      wrapper.update()
      expect(
        wrapper
          .find(`.${SO_PREFIX}-tabs-tab`)
          .at(index)
          .getDOMNode().style.borderColor
      ).toBe(`${pannel[index].border} ${pannel[index].border} ${pannel[index].border} ${pannel[index].border}`)
    })
  })
})

describe('Tabs.Panel[base]', () => {
  test('should set Tabs.Panel base', () => {
    baseTest(Tabs.Panel, `.${SO_PREFIX}-tabs-panel`)
  })
})

describe('Tabs.Panel[disabled]', () => {
  test('should set Tabs.Panel disabled', () => {
    const wrapper = mount(
      <Tabs defaultActive={1}>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
        <Tabs.Panel tab="B">Test</Tabs.Panel>
        <Tabs.Panel tab="C" disabled>
          Test
        </Tabs.Panel>
      </Tabs>
    )
    expect(
      wrapper
        .find(`.${SO_PREFIX}-tabs-tab`)
        .last()
        .hasClass(`${SO_PREFIX}-tabs-disabled`)
    ).toBe(true)
    wrapper
      .find(`.${SO_PREFIX}-tabs-tab`)
      .last()
      .simulate('click')

    expect(
      wrapper
        .find(`.${SO_PREFIX}-tabs-tab`)
        .last()
        .hasClass(`${SO_PREFIX}-tabs-active`)
    ).toBe(false)
  })
})

describe('Tabs.Panel[id]', () => {
  test('should set Tabs.Panel id', () => {
    const pannel = [
      {
        id: 'A',
        tab: 'A',
      },
      {
        id: 'B',
        tab: 'B',
      },
      {
        id: 'C',
        tab: 'C',
      },
    ]

    const handleChange = jest.fn()
    const wrapper = mount(
      <Tabs defaultActive={1} onChange={handleChange}>
        {pannel.map(i => (
          <Tabs.Panel key={i} tab={i.tab} id={i.id}>
            Test
          </Tabs.Panel>
        ))}
      </Tabs>
    )
    wrapper.find(`.${SO_PREFIX}-tabs-tab`).forEach(i => {
      i.simulate('click')
    })

    expect(handleChange).toBeCalledTimes(pannel.length)
    handleChange.mock.calls.forEach((i, index) => {
      expect(i[0]).toBe(pannel[index].id)
    })
  })
})

describe('Tabs.Panel[tab]', () => {
  test('should set Tabs.Panel tab', () => {
    const pannel = [
      {
        id: 'A',
        tab: 'A',
      },
      {
        id: 'B',
        tab: 'B',
      },
      {
        id: 'C',
        tab: 'C',
      },
    ]
    const wrapper = mount(
      <Tabs defaultActive={1}>
        {pannel.map(i => (
          <Tabs.Panel key={i} tab={i.tab} id={i.id}>
            Test
          </Tabs.Panel>
        ))}
      </Tabs>
    )

    wrapper.find(`.${SO_PREFIX}-tabs-tab`).forEach((i, index) => {
      expect(i.text()).toBe(pannel[index].tab)
    })
  })
})

describe('Tabs.Link', () => {
  test('should set Tabs.Link and custom click', () => {
    const handleClick = jest.fn()
    const wrapper = mount(
      <BrowserRouter>
        <Tabs defaultActive={2} shape="line">
          <Tabs.Link>
            <Link to="#tab-link" onClick={handleClick}>
              Link
            </Link>
          </Tabs.Link>
          <Tabs.Link>Link</Tabs.Link>
        </Tabs>
      </BrowserRouter>
    )

    wrapper
      .find('a')
      .first()
      .simulate('click')
    expect(handleClick).toBeCalled()
  })
})

describe('Tabs[unmount]', () => {
  test('should unmount', () => {
    const wrapper = mount(
      <Tabs defaultActive={0}>
        <Tabs.Panel tab="A">Test</Tabs.Panel>
      </Tabs>
    )
    expect(wrapper.find(`.${SO_PREFIX}-tabs`).length).toBe(1)
    wrapper.unmount()
    expect(wrapper.find(`.${SO_PREFIX}-tabs`).length).toBe(0)
  })
})
