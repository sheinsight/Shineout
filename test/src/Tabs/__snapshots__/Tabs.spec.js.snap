// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs[Base] should render correct 1`] = `
<div
  className="so-tabs"
>
  <div
    className="so-tabs-header"
    onClick={[Function]}
  >
    <div
      className="so-tabs-inner"
    >
      <div
        className="so-tabs-scroll"
        style={
          Object {
            "marginLeft": -0,
          }
        }
      >
        <div
          className="so-tabs-tab tab_unique_jv6f42yt"
          onClick={[Function]}
          style={
            Object {
              "background": "transparent",
              "borderColor": "transparent transparent transparent transparent",
              "color": undefined,
            }
          }
        >
          Home
        </div>
        <div
          className="so-tabs-tab so-tabs-active tab_unique_jv6f42yu"
          onClick={[Function]}
          style={
            Object {
              "background": "#fff",
              "borderColor": "#ddd #ddd #fff #ddd",
              "color": "#333",
            }
          }
        >
          Profile
        </div>
        <div
          className="so-tabs-tab tab_unique_jv6f42yv"
          onClick={[Function]}
          style={
            Object {
              "background": "transparent",
              "borderColor": "transparent transparent transparent transparent",
              "color": undefined,
            }
          }
        >
          Contact
        </div>
      </div>
    </div>
    <div
      className="so-tabs-hr"
      style={
        Object {
          "borderColor": "#ddd",
        }
      }
    />
  </div>
  <div
    className="so-tabs-panel so-tabs-show"
    style={
      Object {
        "background": "#fff",
        "color": undefined,
      }
    }
  >
    Profile
  </div>
</div>
`;

exports[`Tabs[Base] should render defined style 1`] = `
<div
  className="so-tabs"
>
  <div
    className="so-tabs-header"
    onClick={[Function]}
  >
    <div
      className="so-tabs-inner"
    >
      <div
        className="so-tabs-scroll"
        style={
          Object {
            "marginLeft": -0,
          }
        }
      >
        <div
          className="so-tabs-tab so-tabs-active tab_unique_jv6f42yw"
          onClick={[Function]}
          style={
            Object {
              "background": "#ffe7ba",
              "borderColor": "transparent transparent #ffe7ba transparent",
              "color": "#333",
            }
          }
        >
          Home
        </div>
        <div
          className="so-tabs-tab tab_unique_jv6f42yx"
          onClick={[Function]}
          style={
            Object {
              "background": "#ffc069",
              "borderColor": "transparent transparent transparent transparent",
              "color": undefined,
            }
          }
        >
          Profile
        </div>
        <div
          className="so-tabs-tab tab_unique_jv6f42yy"
          onClick={[Function]}
          style={
            Object {
              "background": "#d46b08",
              "borderColor": "transparent transparent transparent transparent",
              "color": "#fff",
            }
          }
        >
          <span>
            <i
              className="so-icon so-icon-default fa-user"
              size="default"
              style={
                Object {
                  "fontFamily": "FontAwesome",
                  "fontSize": undefined,
                }
              }
            />
             Contact
          </span>
        </div>
        <div
          className="so-tabs-tab tab_unique_jv6f42yz"
          onClick={[Function]}
          style={
            Object {
              "background": "#873800",
              "borderColor": "transparent transparent transparent transparent",
              "color": "#fff",
            }
          }
        >
          Setting
        </div>
        <div
          className="so-tabs-tab tab_unique_jv6f42z0"
          onClick={[Function]}
          style={
            Object {
              "background": "#f6ffed",
              "borderColor": "#b7eb8f #b7eb8f #b7eb8f #b7eb8f",
              "color": undefined,
            }
          }
        >
          Message
        </div>
      </div>
    </div>
    <div
      className="so-tabs-hr"
      style={
        Object {
          "borderColor": "transparent",
        }
      }
    />
  </div>
  <div
    className="so-tabs-panel so-tabs-show"
    style={
      Object {
        "background": "#ffe7ba",
        "color": undefined,
        "padding": 15,
      }
    }
  >
    Test
  </div>
</div>
`;
